openff/toolkit/topology/molecule.py

```diff
@@ -4538,55 +4598,57 @@ def n_particles(self):
         )
 
     @property
-    def n_atoms(self):
+    def n_atoms(self) -> int:
         """
         The number of Atom objects.
         """
         return len(self._atoms)
 
     @property
-    def n_virtual_sites(self):
+    def n_virtual_sites(self) -> int:
         """
         The number of VirtualSite objects.
         """
         return len(self._virtual_sites)
 
     @property
-    def n_virtual_particles(self):
+    def n_virtual_particles(self) -> int:
         """
         The number of VirtualParticle objects.
         """
         return sum(vsite.n_particles for vsite in self._virtual_sites)
 
     @property
-    def n_bonds(self):
+    def n_bonds(self) -> int:
         """
-        The number of Bond objects.
+        The number of Bond objects in the molecule.
         """
         return sum([1 for bond in self.bonds])
 
     @property
-    def n_angles(self):
-        """int: number of angles in the Molecule."""
+    def n_angles(self) -> int:
+        """Number of angles in the Molecule."""
         self._construct_angles()
         return len(self._angles)
 
     @property
-    def n_propers(self):
-        """int: number of proper torsions in the Molecule."""
+    def n_propers(self) -> int:
+        """Number of proper torsions in the Molecule."""
         self._construct_torsions()
+        assert self._propers is not None
```

Comment:
It's actually to get the function to type check - from inspecting the source `_construct_torsions` always sets `_propers` to a set (though it may be empty). The assert just tells mypy that. I prefer asserts to just silencing the type checker because if something changes you get a crash that should show up in tests. I like asserts as a way to document invariants, but I'd consider any assert that actually raises an exception a bug - it means that an invariant that you're relying on is not being upheld. I'd rather an exception in that case than an incorrect result. I agree that they're not a good way to actually check if something's true, and if a user sees one something's gone wrong.

I'm happy to add a message though :)